stdlib/public/core/StringStorageBridge.swift

```diff
@@ -108,47 +155,33 @@ extension _AbstractStringStorage {
     case .tagged:
       // Tagged means ASCII. If we're equal, our UTF-8 length is the same as our
       // UTF-16 length, so just compare the UTF-8 length (which is faster).
-      otherIsTagged = true
-      fallthrough
+      return _isEqualSlowPath(
+        other: other,
+        otherIsKnownASCII: true,
+        otherIsKnownNSString: false,
+        maybeOtherUTF16Length: nil,
+        maybeOtherASCIIPointer: nil
+      )
+#if arch(arm64)
+    case .constantTagged:
+      let contents = getConstantTaggedCocoaContents(other)!
+      return _isEqualSlowPath(
+        other: other,
+        otherIsKnownASCII: contents.asciiContentsPointer != nil,
+        otherIsKnownNSString: true,
+        maybeOtherUTF16Length: contents.UTF16Length,
+        maybeOtherASCIIPointer: contents.asciiContentsPointer
+      )
```

Comment:
I still think there's an opportunity to simplify the `_isEqualSlowPath`'s interface a bit. Having 2 bools and 2 optionals presents 16 ways to call this function, but not all combinations are possible.

For example, you can order the `constantTagged` case above `.tagged` and drop the `otherIsKnownNSString` parameter, since the fall-through shared by `.tagged` and `.cocoa` will have the `_isNSString()` call. I think you can do something similar for the ASCII-ness based on your earlier comment. Also, it seems like the two maybes are not orthogonal, so that's another factor of 2 you might be able to simplify.

Reply:
I think the new version should make you happier here :)